File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;

    public int[] solve(TreeNode node) {
        if (node == null) {
            return new int[] { 0, 0 };
        }

        int[] left = solve(node.left);
        int[] right = solve(node.right);

        if ((left[1] + right[1] + node.val) / (left[0] + right[0] + 1) == node.val)
            res++;
        // System.out.println( sum +  " " + (1  + left + right));
        return new int[] { 1 + left[0] + right[0], node.val + left[1] + right[1] };
    }

    public int averageOfSubtree(TreeNode root) {
        solve(root);
        return res;
    }
}

File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
   int ans=0;
  
  pair<int,int>solve(TreeNode*root){
     if(root==NULL){
        return {0,0};
     }
     
     pair<int,int> left=solve(root->left);
     pair<int,int> right=solve(root->right);

     int sum=left.first+right.first;
     int cou=left.second+right.second;
     int s=sum+root->val;
     int c=cou+1;

     if(s/c==root->val){
        ans++;
     }
     return {s,c};
    
   }
    int averageOfSubtree(TreeNode* root) {
        
        solve(root);
        
       return ans;

    }
};